src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1837,27 +1763,41 @@ SyncMetadataToNodes(void)
 		return METADATA_SYNC_FAILED_LOCK;
 	}
 
+	List *workerSuccess = NIL;
 	List *workerList = ActivePrimaryNonCoordinatorNodeList(NoLock);
 	WorkerNode *workerNode = NULL;
 	foreach_ptr(workerNode, workerList)
 	{
-		if (workerNode->hasMetadata && !workerNode->metadataSynced)
+		if (!workerNode->hasMetadata)
 		{
-			bool raiseInterrupts = false;
+			/* we have nothing to sync on workers with no metadata */
+			continue;
+		}
 
+		bool setMetadataSynced = true;
+		if (!workerNode->metadataSynced)
+		{
+			bool raiseInterrupts = false;
 			if (!SyncMetadataSnapshotToNode(workerNode, raiseInterrupts))
 			{
 				ereport(WARNING, (errmsg("failed to sync metadata to %s:%d",
 										 workerNode->workerName,
 										 workerNode->workerPort)));
 				result = METADATA_SYNC_FAILED_SYNC;
-			}
-			else
-			{
-				MarkNodeMetadataSynced(workerNode->workerName,
-									   workerNode->workerPort, true);
+				setMetadataSynced = false;
 			}
 		}
+
+		if (setMetadataSynced)
```

Comment:
1) I think the name `workerSuccess` is not descriptive enough, can you update?
2) Do we really need `setMetadataSynced`? Why can't we simply add to `wokerSucces`?
3) we need some comment on what we are doing here.